credoai/evaluators/shap.py

```diff
@@ -0,0 +1,141 @@
+from typing import Dict, List, Optional
+from credoai.evaluators import Evaluator
+from credoai.evidence import TableContainer
+from credoai.evaluators.utils.validation import check_requirements_existence
+from shap import Explainer, Explanation
+from pandas import DataFrame, concat
+from numpy import mean, abs
+
+
+class ShapValues(Evaluator):
+    """
+    This evaluator perform the calculation of shapley values for a dataset/model,
+    leveraging the SHAP package.
+
+    It calculates 2 types of assessments:
+    1. Overall statistics of the shap values across all samples: mean and mean(|x|)
+    2. Optional: individual shapley values for a list of samples
+
+    Parameters
+    ----------
+    samples_ind : Optional[List[int]], optional
+        List of row numbers representing the samples for which to extract individual
+        shapley values , by default None
+    """
+
+    def __init__(self, samples_ind: Optional[List[int]] = None):
+
+        super().__init__()
+        self.samples_ind = samples_ind
+
+    name = "Shap"
+    required_artifacts = ["assessment_data", "model"]
+
+    def _setup(self):
+        self.X = self.assessment_data.X
+        self.model = self.model
+        return self
+
+    def _validate_arguments(self):
+        check_requirements_existence(self)
+
+    def evaluate(self):
+        ## Overall stats
+        self._setup_shap()
+        res = self._get_overall_shap_contributions()
+        self.results = [TableContainer(res, **self.get_container_info())]
+
+        ## Sample specific results
+        if self.samples_ind:
+            labels = {"ordered_feature_names": self.shap_values.feature_names}
+            ind_res = self._get_mult_sample_shapley_values()
+            self.results += [
+                TableContainer(ind_res, **self.get_container_info(labels=labels))
+            ]
+        return self
+
+    def _setup_shap(self):
+        """
+        Setup the explainer given the model and the feature dataset
+        """
+        explainer = Explainer(self.model.predict, self.X)
```

Comment:
It seems that you're just using SHAP's model-agnostic explainer (https://shap.readthedocs.io/en/latest/tabular_examples.html#model-agnostic). If I'm not mistaken the library has different SHAP algorithms for different base models so there might be some computational benefit to actually specifying Tree vs. linear model vs. something else.

Maybe something to track via Lens's model wrappers so that we can use here to better select the SHAP explainer?

Reply:
Agreed. If we can do smart inference based on what's in the `modelinfo` we should use it. Right now `modelinfo` only infers framework.

We can also add an initialization parameter that takes a SHAP explainer. We default to the less-than-optimal model-agnostic explainer otherwise.